// @revolist/revo-dropdown/standalone/revo-list2.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';

const revoListStyleCss = "revo-list{overflow-x:hidden;overflow-y:auto;max-height:100%;display:block}revo-list ul{margin:0;padding:0;outline:0;list-style:none}revo-list ul>li{width:auto;overflow:hidden;font-size:14px;box-sizing:border-box;min-height:48px;font-weight:400;line-height:1.5;padding-top:6px;white-space:nowrap;padding-bottom:6px;padding-left:16px;padding-right:16px;display:flex;position:relative;text-align:left;align-items:center;justify-content:flex-start;text-decoration:none;cursor:pointer}revo-list ul>li.selected{background-color:rgba(0, 0, 0, 0.04)}revo-list ul>li:hover{background-color:rgba(0, 0, 0, 0.04)}";
const RevoListStyle0 = revoListStyleCss;

const RevoDropdownList = /*@__PURE__*/ proxyCustomElement(class RevoDropdownList extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.changed = createEvent(this, "changed", 3);
        this.currentItem = 0;
        this.sourceItems = [];
        this.isFocused = false;
        this.selectedIndex = 0;
        this.template = undefined;
    }
    /** Recived keyboard down from element */
    onKey(e) {
        let item;
        if (!this.isFocused) {
            return;
        }
        switch (e.code) {
            case 'ArrowUp':
                e.preventDefault();
                if (this.currentItem <= 0) {
                    return;
                }
                this.currentItem--;
                break;
            case 'ArrowDown':
                e.preventDefault();
                if (this.sourceItems[this.currentItem + 1]) {
                    this.currentItem++;
                }
                break;
            case 'Tab':
                e.preventDefault();
                item = this.sourceItems[this.currentItem];
                if (item) {
                    this.changed.emit({ item, e });
                }
                break;
            case 'Enter':
                e.preventDefault();
                item = this.sourceItems[this.currentItem];
                if (item) {
                    this.changed.emit({ item, e });
                }
                break;
        }
    }
    async refresh(source) {
        this.sourceItems = source;
    }
    valueChanged(newV) {
        this.currentItem = newV;
    }
    componentWillLoad() {
        this.valueChanged(this.selectedIndex);
    }
    componentDidRender() {
        var _a;
        (_a = this.selectedEl) === null || _a === void 0 ? void 0 : _a.scrollIntoView({
            block: 'nearest',
            inline: 'nearest',
        });
    }
    render() {
        this.selectedEl = undefined;
        const items = [];
        for (let i in this.sourceItems) {
            const item = this.sourceItems[i];
            const isSelected = parseInt(i) === this.currentItem;
            const props = {
                class: { selected: isSelected },
                ref: e => {
                    if (isSelected) {
                        this.selectedEl = e;
                    }
                },
                onClick: e => this.changed.emit({ item, e }),
            };
            const li = h("li", Object.assign({ key: 'cb47c8bde7b1a4229675c49daad83816e03e1461' }, props), this.template(item));
            items.push(li);
        }
        return h("ul", { key: '3a77f6a7bb446d5f380b4f4e774906f59d36e31f' }, items);
    }
    static get watchers() { return {
        "selectedIndex": ["valueChanged"]
    }; }
    static get style() { return RevoListStyle0; }
}, [0, "revo-list", {
        "sourceItems": [16],
        "isFocused": [4, "is-focused"],
        "selectedIndex": [2, "selected-index"],
        "template": [16],
        "currentItem": [32],
        "refresh": [64]
    }, [[4, "keydown", "onKey"]], {
        "selectedIndex": ["valueChanged"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["revo-list"];
    components.forEach(tagName => { switch (tagName) {
        case "revo-list":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, RevoDropdownList);
            }
            break;
    } });
}

export { RevoDropdownList as R, defineCustomElement as d };
