// @revolist/revo-dropdown/dist/collection/components/list/revo-list.js
import { h } from "@stencil/core";
export class RevoDropdownList {
    constructor() {
        this.currentItem = 0;
        this.sourceItems = [];
        this.isFocused = false;
        this.selectedIndex = 0;
        this.template = undefined;
    }
    /** Recived keyboard down from element */
    onKey(e) {
        let item;
        if (!this.isFocused) {
            return;
        }
        switch (e.code) {
            case 'ArrowUp':
                e.preventDefault();
                if (this.currentItem <= 0) {
                    return;
                }
                this.currentItem--;
                break;
            case 'ArrowDown':
                e.preventDefault();
                if (this.sourceItems[this.currentItem + 1]) {
                    this.currentItem++;
                }
                break;
            case 'Tab':
                e.preventDefault();
                item = this.sourceItems[this.currentItem];
                if (item) {
                    this.changed.emit({ item, e });
                }
                break;
            case 'Enter':
                e.preventDefault();
                item = this.sourceItems[this.currentItem];
                if (item) {
                    this.changed.emit({ item, e });
                }
                break;
        }
    }
    async refresh(source) {
        this.sourceItems = source;
    }
    valueChanged(newV) {
        this.currentItem = newV;
    }
    componentWillLoad() {
        this.valueChanged(this.selectedIndex);
    }
    componentDidRender() {
        var _a;
        (_a = this.selectedEl) === null || _a === void 0 ? void 0 : _a.scrollIntoView({
            block: 'nearest',
            inline: 'nearest',
        });
    }
    render() {
        this.selectedEl = undefined;
        const items = [];
        for (let i in this.sourceItems) {
            const item = this.sourceItems[i];
            const isSelected = parseInt(i) === this.currentItem;
            const props = {
                class: { selected: isSelected },
                ref: e => {
                    if (isSelected) {
                        this.selectedEl = e;
                    }
                },
                onClick: e => this.changed.emit({ item, e }),
            };
            const li = h("li", Object.assign({ key: 'cb47c8bde7b1a4229675c49daad83816e03e1461' }, props), this.template(item));
            items.push(li);
        }
        return h("ul", { key: '3a77f6a7bb446d5f380b4f4e774906f59d36e31f' }, items);
    }
    static get is() { return "revo-list"; }
    static get originalStyleUrls() {
        return {
            "$": ["revo-list.style.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["revo-list.style.css"]
        };
    }
    static get properties() {
        return {
            "sourceItems": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "any[]",
                    "resolved": "any[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Define object mapping for id/value"
                },
                "defaultValue": "[]"
            },
            "isFocused": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "is-focused",
                "reflect": false,
                "defaultValue": "false"
            },
            "selectedIndex": {
                "type": "number",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Selected Value Index"
                },
                "attribute": "selected-index",
                "reflect": false,
                "defaultValue": "0"
            },
            "template": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "(item: any) => VNode",
                    "resolved": "(item: any) => VNode",
                    "references": {
                        "VNode": {
                            "location": "import",
                            "path": "@stencil/core",
                            "id": "node_modules::VNode"
                        }
                    }
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            }
        };
    }
    static get states() {
        return {
            "currentItem": {}
        };
    }
    static get events() {
        return [{
                "method": "changed",
                "name": "changed",
                "bubbles": false,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "{ item: any; e: any }",
                    "resolved": "{ item: any; e: any; }",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "refresh": {
                "complexType": {
                    "signature": "(source: any[]) => Promise<void>",
                    "parameters": [{
                            "name": "source",
                            "type": "any[]",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "",
                    "tags": []
                }
            }
        };
    }
    static get watchers() {
        return [{
                "propName": "selectedIndex",
                "methodName": "valueChanged"
            }];
    }
    static get listeners() {
        return [{
                "name": "keydown",
                "method": "onKey",
                "target": "document",
                "capture": false,
                "passive": false
            }];
    }
}
