// @revolist/revogrid/standalone/revogr-temp-range2.js
/*!
 * Built by Revolist OU ❤️
 */
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { t as throttle } from './throttle.js';
import { T as TMP_SELECTION_BG_CLASS } from './consts.js';
import { a as getCell, s as styleByCellProps } from './selection.utils.js';

const revogrTempRangeStyleCss = ".temp-bg-range{display:block !important;position:absolute;pointer-events:none;z-index:9;border:1px solid rgb(255, 94, 0);box-sizing:border-box}.temp-bg-range.Selection{border:1px dashed gray}.temp-bg-range>div{width:1px;height:1px;position:absolute}.temp-bg-range>div.top{top:-1px}.temp-bg-range>div.bottom{bottom:-1px}.temp-bg-range>div.left{left:-1px}.temp-bg-range>div.right{right:-1px}";

const RevogrFocus = /*@__PURE__*/ proxyCustomElement(class RevogrFocus extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.onChange = throttle((e) => this.doChange(e), 300);
    }
    doChange(e) {
        e === null || e === void 0 ? void 0 : e.scrollIntoView({
            block: 'nearest',
            inline: 'nearest',
        });
    }
    componentDidRender() {
        if (this.el) {
            this.onChange(this.el);
        }
    }
    render() {
        const data = this.selectionStore.get('tempRange');
        const type = this.selectionStore.get('tempRangeType');
        if (!data) {
            return;
        }
        let directionY = 'bottom';
        let derectionX = 'right';
        const range = this.getRange();
        if (!range) {
            return;
        }
        if (data.y < range.y) {
            directionY = 'top';
        }
        if (data.x < range.x) {
            derectionX = 'left';
        }
        const directionClass = `${derectionX} ${directionY}`;
        const cell = getCell(data, this.dimensionRow.state, this.dimensionCol.state);
        const styles = styleByCellProps(cell);
        const props = {
            class: {
                [TMP_SELECTION_BG_CLASS]: true,
                [type || '']: true,
            },
            style: styles,
            hidden: false
        };
        return (h(Host, Object.assign({}, props), h("div", { class: directionClass, ref: (e) => (this.el = e) })));
    }
    getRange() {
        const range = this.selectionStore.get('range');
        if (range) {
            return range;
        }
        const focus = this.selectionStore.get('focus');
        if (!focus) {
            return null;
        }
        return Object.assign(Object.assign({}, focus), { x1: focus.x, y1: focus.y });
    }
    static get style() { return revogrTempRangeStyleCss; }
}, [256, "revogr-temp-range", {
        "selectionStore": [16, "selection-store"],
        "dimensionRow": [16, "dimension-row"],
        "dimensionCol": [16, "dimension-col"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["revogr-temp-range"];
    components.forEach(tagName => { switch (tagName) {
        case "revogr-temp-range":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, RevogrFocus);
            }
            break;
    } });
}

export { RevogrFocus as R, defineCustomElement as d };
